WebAPI: /v1/objects/:id/children added

alkk opened the pull request.

alkk committed:
WebAPI: /v1/objects/:id/children added

2128506 commented on src/server/webapi/objects.cpp:
```diff
@@ -224,6 +224,149 @@ int H_ObjectDetails(Context *context)
+
+   visited.insert(objectId);
+
+   unique_ptr<SharedObjectArray<NetObj>> children = g_idxObjectById.getObjects(
```
no need to scan all objects to find direct parents - they are accessible directly from object itself

2128506 replied:
no need to scan all objects to find direct parents - they are accessible directly from object itself

2128506 commented on src/server/webapi/objects.cpp:
```diff
@@ -224,6 +224,149 @@ int H_ObjectDetails(Context *context)
+   }
+
+   // Get maximum depth parameter (default 10)
+   int maxDepth = 10;
```
can use getQueryParameterAsInt32 with default value:
int maxDepth = context->getQueryParameterAsInt32("maxDepth", 10);

2128506 commented on src/server/webapi/main.cpp:
```diff
@@ -261,6 +262,9 @@ static bool InitModule(Config *config)
    RouteBuilder("v1/objects/:object-id/take-screenshot")
       .GET(H_TakeScreenshot)
       .build();
+   RouteBuilder("v1/objects/:object-id/children")
```
I would call it sub-tree instead of children to better represent meaning

2128506 commented on src/server/webapi/objects.cpp:
```diff
@@ -224,6 +224,149 @@ int H_ObjectDetails(Context *context)
+
+   visited.insert(objectId);
+
+   unique_ptr<SharedObjectArray<NetObj>> children = g_idxObjectById.getObjects(
```
can be replaced by loop on object parents

alkk committed:
WebAPI: address review comments for object sub-tree endpoint

- Made helper functions static
- Replaced object scanning with direct parent-child access using object->getChildren()
- Used getQueryParameterAsInt32 for maxDepth parameter parsing
- Renamed endpoint from /v1/objects/:id/children to /v1/objects/:id/sub-tree
- Renamed handler function from H_ObjectChildren to H_ObjectSubTree

2128506 commented on src/server/webapi/objects.cpp:
```diff
@@ -227,27 +227,30 @@ int H_ObjectDetails(Context *context)
-            return false;
-         return obj->isDirectParent(objectId);
-      });
+   shared_ptr<NetObj> object = FindObjectById(objectId);
```
caller of this function already has pointer to object. It can be passed as argument instead of passing ID and finding object.

2128506 commented on src/server/webapi/objects.cpp:
```diff
@@ -267,37 +270,41 @@ bool HasDescendantsMatchingClassFilter(uint32_t objectId, uint32_t userId, const
-      {
-         if (obj->isHidden() || obj->isSystem() || obj->isDeleted() || !obj->checkAccessRights(userId, OBJECT_ACCESS_READ))
-            return false;
+   shared_ptr<NetObj> object = FindObjectById(objectId);
```
caller of this function already has pointer to object. It can be passed as argument instead of passing ID and finding object.

alkk committed:
WebAPI: optimize object sub-tree functions by passing object pointers

Pass shared_ptr<NetObj> directly to HasDescendantsMatchingClassFilter and
BuildNestedObjectTree functions instead of object IDs to avoid redundant
FindObjectById lookups and improve performance.

alkk committed:
WebAPI: optimize object sub-tree functions by using shared pointers directly

Replace FindObjectById() calls with direct usage of children->getShared(i)
to improve performance and eliminate unnecessary object lookups.
